fix: `<Link href="../">`

trin4ik opened the pull request:
at now `<Link href="../">` didnt work, because [/src/link/useLinkToPath.ts](https://github.com/expo/router/blob/12a53bd2919ae81dd52ead389ca239020416e3cb/packages/expo-router/src/link/useLinkToPath.ts#L22) has a strong comparison:
```javascript
if (!to.startsWith("/")) {
  if (/:\/\//.test(to)) {
    // Open external link
    Linking.openURL(to);
    return;
  } else {
    throw new Error(
      `The href must start with '/' (${to}) or be a fully qualified URL.`
    );
  }
}
```
and if we try `<Link href="../" />`, we got exception `The href must start with '/' (${to}) or be a fully qualified URL.`

### suggestions:
* attach relative path to the current. 
```javascript
if (to.startsWith("../") || to.startsWith("./")) {
  to = current + "/" + to;
}
```
*  and normalize all links (non-external, of course). as a last resort, the user will get a 404 but not an exception.

for convenience i add `useCurrentRoute` with some sugar, like add first `/` to the path, if not. this is for if we start app from deep link, like `exp://127.0.0.1:19000/--/test`, by default `getCurrentRoute().path` will be `test`, not `/test`. `useCurrentRoute` always return current path like `/test` or `/` or `/profile?name=smith`.

`normalizePath` method i put to `/src/link/href.ts`, mb he doesn't belong there.

EvanBacon commented on packages/expo-router/src/link/useLinkToPath.ts:
```diff
@@ -25,9 +29,12 @@ export function useLinkToPath() {
-          throw new Error(
-            `The href must start with '/' (${to}) or be a fully qualified URL.`
-          );
+          // if relative, need append to current
```
This whole statement could be rewritten to be a bit nicer if (_matches url_) else normalize path.

EvanBacon commented on packages/expo-router/src/link/href.ts:
```diff
@@ -42,3 +42,14 @@ function createQuery(query: Record<string, any>) {
     .map((key) => `${key}=${query[key]}`)
     .join("&");
 }
+export function normalizePath(path: string): string {
```
needs tests

trin4ik replied:
ok, trying do it tomorrow

EvanBacon commented on packages/expo-router/src/useCurrentRoute.ts:
```diff
@@ -18,3 +18,18 @@ export function useRootNavigation() {
+      "useCurrentRoute must be used within a NavigationContainerContext"
+    );
+  }
+  const root = useRootNavigation();
```
this assertion also runs twice.

trin4ik replied:
I will fix this

EvanBacon commented on packages/expo-router/src/useCurrentRoute.ts:
```diff
@@ -18,3 +18,18 @@ export function useRootNavigation() {
+    );
+  }
+  const root = useRootNavigation();
+  let path = root?.getCurrentRoute()?.path;
```
In my testing, this is mostly not working. So far looks like at least replace doesn't support this.

trin4ik commented:
ok, i added test for `normalizePath` and fixed `useCurrentRoute`, but it is still draft. it is because `replace`, agrrr.

* i think we need `useCurrentRoute` not from `root.getCurrentRoute()`, because if we `link.replace('/foo/bar)`, root.getCurrentRoute() will has null path. i think we need get current route with `stateToPath` way. now we have `getPathFromState` method, but with `link.replace('/foo/bar')` and structure like
```
app
├── (root)
│   ├── (app)
│   │   ├── [...bar]
│   │   │   ├── index.js
│   │   │   └── orders.js
│   │   ├── [...bar].tsx
│   │   ├── [foo].tsx
│   │   └── index.tsx
│   └── sign-in.tsx
└── (root).tsx
```
we got `/(root)/(app)?bar=foo&bar=bar`, seems we need method to convert this path to url, like `/(root)/(app)?bar=foo&bar=bar -> /foo/bar`

* replace is bottleneck. i used replace with react-navigation for replace navigtion stack auth/deauth in my apps. and it work great. but if we use replace with file-based router, what we want? let's say, if we `link.replace('/foo/bar')`, what current path we want to see? `/` or `/foo/bar`? in web we havent replace, because user can rewrite url by him self. in file-system, what is replace path? like jail? it's definitely not symlink, it is like jail, i think. jail, where you cant go back after replace. after replace your current path root be `/`, but for app it is `/foo/bar/`, right? and if after `link.replace('/foo/bar')` you go to `link.push('/test')`, yout current path be `/test`, but for router it will be `/foo/bar/test`, right? in this case what happend, if you try `link.replace('/')`?

I think we need to decide if we need `replace` and if so, in what form and how it will relate to file-system-base router.
and I'm not trying to be clever or confuse you, I really like this project, I just want to help.

EvanBacon commented:
Seems resolved in #73
